### DISCUSSION TITLE
In-place compaction with decoupled lookback

### DISCUSSION THREAD

**Discussion by kevkrist** [NVIDIA/cccl#1684] [CUB] [answered]
Hi!
I'm wondering if it should be easy to set up my own kernel with decoupled lookback. Is there an example? How might I begin? I'd love to use as much CUB as possible! For context, I essentially want to reduce by key, but the key and the value are both extracted from a 64-bit hash, which I'd like to extract in-place, and then I want to use the key as the write index into an output array with the reduced value, for a specialized in-place write. The engineering effort required over a chained scan is pretty steep, I feel, so it's unlikely I'll implement my own decoupled lookback, and it'd be REALLY NICE for my research to have a way of using decoupled lookback in general for my own specialized compactions. Thanks!

**Comment by jrhemstad** [comment, marked as answer]
Hey @kevkrist , you are in luck! 

We have an example demonstrating the machinery that CUB exposes for decoupled look back that you can check out here: https://github.com/NVIDIA/cccl/blob/main/cub/examples/device/example_device_decoupled_look_back.cu

**Reply by kevkrist** [reply]
Amazing thank you!
